File: rust-core/src/filter.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FilteredItem {
    pub source: String,
    pub repo: String,
    pub item_type: String,
    pub title: String,
    pub priority: Priority,
    pub reason: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Priority {
    Urgent,
    Important,
    Noise,
}

impl Priority {
    pub fn as_str(&self) -> &'static str {
        match self {
            Priority::Urgent => "urgent",
            Priority::Important => "important",
            Priority::Noise => "noise",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Notification {
    pub source: String,
    pub repo: String,
    pub title: String,
    pub ntype: String,
    pub reason: String,
}

pub fn classify(notif: &Notification) -> Priority {
    match notif.ntype.as_str() {
        "PullRequest" => match notif.reason.as_str() {
            "review_requested" | "mention" | "author" => Priority::Urgent,
            _ => Priority::Important,
        },
        "Issue" => match notif.reason.as_str() {
            "assign" | "mention" => Priority::Urgent,
            "author" => Priority::Important,
            _ => Priority::Important,
        },
        "Discussion" => match notif.reason.as_str() {
            "mention" => Priority::Urgent,
            _ => Priority::Important,
        },
        "CheckSuite" | "CheckRun" | "WorkflowRun" => Priority::Noise,
        "Release" => Priority::Important,
        _ => Priority::Noise,
    }
}

pub fn filter_all(notifs: &[Notification]) -> (Vec<FilteredItem>, Vec<FilteredItem>, usize) {
    let mut urgent = Vec::new();
    let mut important = Vec::new();
    let mut noise_count = 0;

    for n in notifs {
        let priority = classify(n);
        let item = FilteredItem {
            source: n.source.clone(),
            repo: n.repo.clone(),
            item_type: n.ntype.clone(),
            title: n.title.clone(),
            priority,
            reason: reason_text(&n.reason, &n.ntype),
        };

        match priority {
            Priority::Urgent => urgent.push(item),
            Priority::Important => important.push(item),
            Priority::Noise => noise_count += 1,
        }
    }

    (urgent, important, noise_count)
}

fn reason_text(reason: &str, ntype: &str) -> String {
    match ntype {
        "PullRequest" => match reason {
            "review_requested" | "mention" => "you were mentioned or asked to review".to_string(),
            "author" => "your PR has activity".to_string(),
            _ => "pr activity".to_string(),
        },
        "Issue" => match reason {
            "assign" | "mention" => "you were assigned or mentioned".to_string(),
            "author" => "your issue has activity".to_string(),
            _ => "issue activity".to_string(),
        },
        "Discussion" => match reason {
            "mention" => "you were mentioned".to_string(),
            _ => "discussion activity".to_string(),
        },
        "CheckSuite" | "CheckRun" | "WorkflowRun" => "ci status".to_string(),
        "Release" => "new release".to_string(),
        _ => reason.to_string(),
    }
}

pub fn group_by_repo(items: &[FilteredItem]) -> HashMap<String, Vec<&FilteredItem>> {
    let mut groups: HashMap<String, Vec<&FilteredItem>> = HashMap::new();
    for item in items {
        groups.entry(item.repo.clone()).or_default().push(item);
    }
    groups
}

File: internal/agent/filter.go
package agent

import (
	"fmt"
	"strings"

	"github.com/valtors/pulse/internal/connect"
)

type Priority int

const (
	PriorityUrgent Priority = iota
	PriorityImportant
	PriorityNoise
)

type FilteredItem struct {
	Source   string `json:"source"`
	Repo     string `json:"repo"`
	Type     string `json:"type"`
	Title    string `json:"title"`
	Priority string `json:"priority"`
	Reason   string `json:"reason"`
}

func priorityName(p Priority) string {
	switch p {
	case PriorityUrgent:
		return "urgent"
	case PriorityImportant:
		return "important"
	default:
		return "noise"
	}
}

func filterGitHub(notifs []connect.GitHubNotification) []FilteredItem {
	var items []FilteredItem
	for _, n := range notifs {
		item := FilteredItem{
			Source: "github",
			Repo:   n.Repository.FullName,
			Type:   n.Subject.Type,
			Title:  n.Subject.Title,
		}

		switch n.Subject.Type {
		case "PullRequest":
			if n.Reason == "review_requested" || n.Reason == "mention" {
				item.Priority = priorityName(PriorityUrgent)
				item.Reason = "you were mentioned or asked to review"
			} else if n.Reason == "author" {
				item.Priority = priorityName(PriorityUrgent)
				item.Reason = "your PR has activity"
			} else {
				item.Priority = priorityName(PriorityImportant)
				item.Reason = "pr activity"
			}
		case "Issue":
			if n.Reason == "assign" || n.Reason == "mention" {
				item.Priority = priorityName(PriorityUrgent)
				item.Reason = "you were assigned or mentioned"
			} else if n.Reason == "author" {
				item.Priority = priorityName(PriorityImportant)
				item.Reason = "your issue has activity"
			} else {
				item.Priority = priorityName(PriorityImportant)
				item.Reason = "issue activity"
			}
		case "Discussion":
			if n.Reason == "mention" {
				item.Priority = priorityName(PriorityUrgent)
				item.Reason = "you were mentioned"
			} else {
				item.Priority = priorityName(PriorityImportant)
				item.Reason = "discussion activity"
			}
		case "CheckSuite":
			item.Priority = priorityName(PriorityNoise)
			item.Reason = "ci status"
		case "Release":
			item.Priority = priorityName(PriorityImportant)
			item.Reason = "new release"
		default:
			item.Priority = priorityName(PriorityNoise)
			item.Reason = n.Reason
		}

		items = append(items, item)
	}
	return items
}

func groupByRepo(items []FilteredItem) map[string][]FilteredItem {
	groups := make(map[string][]FilteredItem)
	for _, item := range items {
		groups[item.Repo] = append(groups[item.Repo], item)
	}
	return groups
}

func countPriority(items []FilteredItem) map[string]int {
	counts := map[string]int{"urgent": 0, "important": 0, "noise": 0}
	for _, item := range items {
		counts[item.Priority]++
	}
	return counts
}

func formatFiltered(items []FilteredItem) string {
	if len(items) == 0 {
		return "nothing to report. you're caught up."
	}

	counts := countPriority(items)
	var sections []string

	if counts["urgent"] > 0 {
		sections = append(sections, formatPriority(items, PriorityUrgent, "URGENT"))
	}
	if counts["important"] > 0 {
		sections = append(sections, formatPriority(items, PriorityImportant, "NEEDS ATTENTION"))
	}

	noiseCount := counts["noise"]
	if noiseCount > 0 {
		sections = append(sections, fmt.Sprintf("\nNOISE: %d items filtered (ci failures, status checks, automated noise)", noiseCount))
	}

	return strings.Join(sections, "\n")
}

func formatPriority(items []FilteredItem, p Priority, label string) string {
	var lines []string
	lines = append(lines, fmt.Sprintf("\n%s:", label))
	groups := groupByRepo(items)
	for repo, repoItems := range groups {
		if !hasPriority(repoItems, p) {
			continue
		}
		lines = append(lines, fmt.Sprintf("  %s", repo))
		for _, item := range repoItems {
			if item.Priority != priorityName(p) {
				continue
			}
			lines = append(lines, fmt.Sprintf("    [%s] %s - %s", item.Type, item.Title, item.Reason))
		}
	}
	return strings.Join(lines, "\n")
}

func hasPriority(items []FilteredItem, p Priority) bool {
	for _, item := range items {
		if item.Priority == priorityName(p) {
			return true
		}
	}
	return false
}
